pkg/agent/provisioning.go

```diff
@@ -34,46 +28,118 @@ func createOrUpdateCredentialRequestDefinition(data *v1alpha1.CredentialRequestD
+	thisCred := &crdBuilderOptions{
+		provProps: make([]provisioning.PropertyBuilder, 0),
+		reqProps:  make([]provisioning.PropertyBuilder, 0),
+	}
+	for _, o := range options {
+		o(thisCred)
+	}
+
+	provSchema := provisioning.NewSchemaBuilder()
+	for _, provProp := range thisCred.provProps {
+		provSchema.AddProperty(provProp)
+	}
+
+	reqSchema := provisioning.NewSchemaBuilder()
+	for _, props := range thisCred.reqProps {
+		reqSchema.AddProperty(props)
+	}
+
+	return provisioning.NewCRDBuilder(createOrUpdateCredentialRequestDefinition).
+		SetName(thisCred.name).
+		SetProvisionSchema(provSchema).
+		SetRequestSchema(reqSchema)
+}
+
+// withCRDName - set another name for the CRD
+func withCRDName(name string) func(c *crdBuilderOptions) {
+	return func(c *crdBuilderOptions) {
+		c.name = name
+	}
+}
+
+// WithCRDProvisionSchemaProperty - add more provisioning properties
+func WithCRDProvisionSchemaProperty(prop provisioning.PropertyBuilder) func(c *crdBuilderOptions) {
+	return func(c *crdBuilderOptions) {
+		c.provProps = append(c.provProps, prop)
+	}
+}
+
+// WithCRDRequestSchemaProperty - add more request properties
+func WithCRDRequestSchemaProperty(prop provisioning.PropertyBuilder) func(c *crdBuilderOptions) {
+	return func(c *crdBuilderOptions) {
+		c.reqProps = append(c.reqProps, prop)
+	}
+}
+
+// WithCRDOAuthSecret - set that the Oauth cred is secret based
+func WithCRDOAuthSecret() func(c *crdBuilderOptions) {
+	return func(c *crdBuilderOptions) {
+		c.name = provisioning.OAuthSecretCRD
+		c.provProps = append(c.provProps,
+			provisioning.NewSchemaPropertyBuilder().
+				SetName(provisioning.OauthClientSecret).
+				SetLabel("Client Secret").
+				SetRequired().
+				IsString().
+				IsEncrypted())
+	}
+}
+
+// WithCRDOAuthPublicKey - set that the Oauth cred is key based
+func WithCRDOAuthPublicKey() func(c *crdBuilderOptions) {
+	return func(c *crdBuilderOptions) {
+		c.name = provisioning.OAuthPublicKeyCRD
+		c.reqProps = append(c.reqProps,
+			provisioning.NewSchemaPropertyBuilder().
+				SetName(provisioning.OauthPublicKey).
+				SetLabel("Public Key").
+				SetRequired().
+				IsString())
+	}
 }
 
 // NewAPIKeyCredentialRequestBuilder - add api key base properties for provisioning schema
-func NewAPIKeyCredentialRequestBuilder() provisioning.CredentialRequestBuilder {
-	if _, err := agent.cacheManager.GetAccessRequestDefinitionByName(apikeyARD); err != nil {
-		NewAccessRequestBuilder().SetName(apikeyARD).Register()
-	}
-	return NewCredentialRequestBuilder().
-		SetName(apikeyCRD).
-		SetProvisionSchema(provisioning.NewSchemaBuilder().
-			AddProperty(
-				provisioning.NewSchemaPropertyBuilder().
-					SetName("key").
-					SetLabel("API Key").
-					SetRequired().
-					IsString().
-					IsEncrypted()))
+func NewAPIKeyCredentialRequestBuilder(options ...func(*crdBuilderOptions)) provisioning.CredentialRequestBuilder {
```

Comment:
Do agents need to be updated?

Reply:
Unless you need to add options, not for this call.